crates/futures/src/lib.rs

```diff
@@ -289,9 +316,108 @@ fn _future_to_promise(future: Box<dyn Future<Item = JsValue, Error = JsValue>>)
         Waiting(Arc<Package>),
     }
 
+    #[cfg(target_feature = "atomics")]
+    struct Waker {
+        array: Vec<i32>,
+        notified: AtomicBool,
+    };
+
+    #[cfg(target_feature = "atomics")]
+    impl Waker {
+        fn new(array: Vec<i32>, notified: bool) -> Self {
+            Waker {
+                array,
+                notified: AtomicBool::new(notified),
+            }
+        }
+    }
+
+    #[cfg(target_feature = "atomics")]
+    impl Notify for Waker {
+        fn notify(&self, id: usize) {
+            console_log!("Waker notify");
+            if !self.notified.swap(true, Ordering::SeqCst) {
+                console_log!("Waker, inside if");
+                let memory_buffer = wasm_bindgen::memory()
+                .dyn_into::<WebAssembly::Memory>()
+                .expect("Should cast a memory to WebAssembly::Memory")
+                .buffer();
+
+                let array_location = self.array.as_ptr() as u32 / 4;
+                let array = Int32Array::new(&memory_buffer)
+                    .subarray(array_location, array_location + self.array.len() as u32);
+
+                let _ = Atomics::notify(&array, id as u32);
+            }
+        }
+    }
+
+    #[cfg(target_feature = "atomics")]
+    fn poll_again(package: Arc<Package>, id: usize) {
+        console_log!("poll_again called");
+        let me = match package.notified.replace(State::Notified) {
```

Comment:
This `notified` check will want to be the same as the `notified` field in `Waker` to ensure all the races align well

Reply:
I don't get it yet. Adding the following check after the `match` statement fixes performance problems, but breaks the whole thing
```rust
if package.waker.notified.swap(true, Ordering::SeqCst) {
    return;
}
```